programs/stake/src/stake_state.rs

```diff
@@ -560,6 +569,15 @@ impl Stake {
             Ok(())
         }
     }
+
+    fn reduce_overage(&mut self, balance: u64, rent_exempt_balance: u64) {
+        // at this moment, balance >>> rent_excempt_balance (no overflow risk)
+        //   => no there could be possibility of non-rent-exempt stake accounts
+        let correct_stake = balance.saturating_sub(rent_exempt_balance);
+        if self.delegation.stake > correct_stake {
+            self.delegation.stake = correct_stake;
+        }
```

Comment:
note that we can't the opposite thing of rewriting to stake more which has miss to be redelegated. We can't detect the good order (deposit/undelegate => delegate) from bad order (undelegate => delegate => deposit).

Reply:
yeah, I've changed the behavior here.